Intragenic deletions
--------------------

Arriba can detect intragenic inversions and duplications, but not deletions. This is because deletions within a gene are difficult to distinguish from ordinary splicing in RNA-Seq data. Arriba's statistical model to find significant events is not applicable to the identification of a significant lack of exon coverage. These questions are better answered by indel callers, whole-genome sequencing, or algorithms to identify differential exon expression. For these reasons, Arriba does not report any intragenic deletions.

RefSeq annotation
-----------------

It is recommended to use annotation from GENCODE or ENSEMBL to run Arriba. RefSeq annotation has less comprehensive annotation of splice sites. Moreover, RefSeq does not annotate the immunoglobulin/T-cell receptor loci. These shortcomings reduce the sensitivity of fusion detection. Users who want to use RefSeq nonetheless are advised to copy the immunoglobulin/T-cell receptor loci annotation from GENCODE/ENSEMBL as a workaround.

Memory consumption
------------------

Arriba usually consumes less than 10 GB of RAM. Approximately 1 GB of RAM is consumed per million chimeric read pairs, plus 4 GB of static overhead to load the assembly and gene annotation. Particularly multiple myeloma samples frequently exceed the normal memory requirements due to countless rearrangements in the immunoglobulin loci. In order to reduce the memory footprint, Arriba can be instructed to subsample reads when an event has a sufficient number of supporting reads. By default, further reads are ignored, once an event has reached 300 supporting reads (see parameter `-U`). Arriba issues a `WARNING: some fusions were subsampled, because they have more than 300 supporting reads` when this threshold has been hit.

However, excessive memory consumption can indicate a user error. So before reducing the maximum number of supporting reads, users should carefully check their scripts/data for mistakes. For example, if paired-end FastQ files are mistakenly passed to STAR in the wrong order, STAR will align almost all reads as discordant mates. Similarly, if the reads in paired-end FastQ files are not ordered properly (i.e., collated by name), then most of them will be aligned in a discordant fashion. When Arriba consumes an unusual amount of memory, users should interrogate the file `Log.final.out` of STAR. If the `% of chimeric reads` reported in the log file is high, then scripts and input files should be checked for errors. The `% of chimeric reads` is normally in the range of 1-10%, with the exception of very few cancer types (such as multiple myeloma), where they are often much higher.

Adapter trimming
----------------

In most cases, it is not necessary to trim adapters, because STAR is capable of performing clipped alignments, such that reads with adapters will simply be aligned partially. For this reason, the demo script `run_arriba.sh` does not perform adapter trimming. However, if many reads contain a substantial fraction of adapter sequence, it can be beneficial to remove adapters for improved sensitivity, because STAR dismisses chimeric alignments when a big fraction of a read cannot be aligned (as controlled by the parameter `--outFilterMatchNminOverLread`). To this end, the STAR parameter `--clip3pAdapterSeq` can be used or a specialized adapter trimming tool.

Small insert size
-----------------

Due to a [bug in STAR](https://github.com/alexdobin/STAR/issues/1135), when the insert size is small, such that paired-end mates fully overlap, STAR may produce malformed chimeric alignments. This is a bug in the paired-end overlap merge feature of STAR (see parameter `--peOverlapNbasesMin`) and only manifests when this feature is enabled. Despite this bug, it is highly recommended to enable the feature, because otherwise chimeric reads of fully (or partially) overlapping paired-end mates will not be aligned at all. The bug only affects fully overlapping paired-end mates (but not partially overlapping mates) and of those only 50% at random, such that even in view of the bug, there still is a major benefit in enabling the paired-end overlap merge feature. However, when the majority of the reads overlaps fully, this means that up to 50% of the chimeric alignments cannot be used by Arriba, which obviously impairs fusion detection. In such cases, it is recommended to align just one of the mates of fully overlapping reads in single-end mode, where the bug does not apply. More importantly, users should seek to improve their library preparation protocol to produce longer insert sizes, because there is no point in performing paired-end sequencing when the information of the paired-end mates is identical. This is effectively the same as performing single-end sequencing. For optimal fusion detection, the insert size should be just about big enough that the paired-end mates do not overlap.

Viral detection
---------------

Detection of virus expression and viral integration sites is limited to the viruses that are built into the STAR index and assembly used to run Arriba. When the script `download_references.sh` was used to generate the reference files, all ~12,000 RefSeq virus genomes as well as ~4,500 human-infecting virus strains related to the RefSeq viruses are included into the reference files. This includes all common cancer-associated viruses, such as human papillomavirus, Merkel cell polyomavirus, EBV, HIV, Hepatitis B/C virus, HTLV-1, and others. Users who wish to detect viruses beyond this default set need to append the respective virus genome sequence to the assembly and build the STAR index themselves.

Not all viruses integrate into the human genome. Their presence can often be detected nonetheless by checking if reads align to the respective virus genome. Arriba comes with a [script `quantify_virus_expression.sh`](utility-scripts.md#quantify-virus-expression), which counts reads mapping to virus genomes and applies a few filters to remove false positives.

Due to sequence similarity between related virus strains, it happens occasionally that STAR maps reads to a related virus strain instead of to the strain that actually infected a given sample. Arriba then sometimes reports integrations of multiple related strains in a single sample. Although it is usually wrong of Arriba to report that a single sample harbors integrations of multiple strains, the positions of the various integration sites can very well be correct, and therefore all of them are reported - even if their alignments come from related but distinct strains. Similarly, when a cell line was genetically engineered using cloning vectors, the vector may sometimes be detected by Arriba as being integrated near the engineered gene (and possibly other places in the genome in case of off-target integration). Since cloning vectors as such are not part of the RefSeq database, Arriba usually reports integration of a virus with similar sequence to the cloning vector. For example, cloning vectors derived from adeno-associated viruses are frequently reported as the woodchuck hepatitis virus.

Targeted sequencing
-------------------

Arriba can be used to analyze RNA sequencing data from a panel of genes. However, Arriba was primarily developed for untargeted/unbiased sequencing. Focal, very deep sequencing depths as seen with targeted sequencing do not match Arriba's statistical model of the expected coverage distribution. Estimation of the background noise does not work as well here, which means Arriba will likely produce more false positives. Low-confidence fusions should probably be ignored altogether. Possibly, one should additionally remove fusions with a lower expression than 1% of the coverage, i.e., the sum of `split_reads1`, `split_reads2`, and `discordant_mates` should be at least 1% of the sum of `coverage1` and `coverage2`.

Moreover, one should be aware that the panel only covers cancer-relevant genes. So every reported fusion prediction will affect a potentially interesting gene. Common sense should be applied whether the fusion actually makes sense in the given cancer type. For example, some panels include the fusion SLC45A3-ELK4. This fusion is primarily relevant to prostate cancer. It would be unusual to observe it in the context of other cancer types. Yet it can be detected in many samples, because it is a read-through fusion that is also seen in benign tissue and because targeted sequencing over-amplifies even slightly expressed transcripts. Generally speaking, the false positives induced by targeted sequencing mostly affect neighboring genes, because the close proximity of genes gives rise to random cis-/trans-spliced transcripts, which are amplified by the panel to higher levels than one would expect with untargeted RNA-Seq. So fusion predictions involving genes in proximity should be treated with caution. More advice on identifying potential false positives and evaluating the relevance of predicted fusions can be found in the chapter [Interpretation of results](interpretation-of-results.md).

Lastly, it is recommended to use the parameter `-u` to enable external duplicate marking instead of using Arriba's internal duplicate marking method. For targeted sequencing, it is common practice to tag fragments with unique molecular identifiers (UMIs) prior to PCR amplification and sequencing. Arriba is oblivious to UMIs. It marks duplicates solely based on the fact that reads have identical mapping coordinates. So it may discard reads with distinct UMIs as PCR duplicates. It is better to perform duplicate marking externally based on the UMIs prior to running Arriba. The parameter `-u` instructs Arriba to not perform its own detection of duplicates and instead rely on duplicates being marked as such in the BAM file using the flag `BAM_FDUP`.

